### DISCUSSION TITLE
n-th root for negative values, for odd n

### DISCUSSION THREAD

**Discussion by romanszewczyk** [MilesCranmer/PySR#434] [Q&A] [answered]
Hi!

I am trying to add to regressor, a custom function calculating 5th root, 7th root and 11th root. 

I tried the following commands:

_unary_operators=["sqrt", "square", "cbrt", "cube", 
                            "inv(x)=1/x",
                            "nroot5(x)=x^(1/5)","nroot7(x)=x^(1/7)","nroot11(x)=x^(1/11)"],_

_extra_sympy_mappings={"inv": lambda x: 1 / x, 
                          "nroot5": lambda x: sympy.Pow(x,1/5), 
                          "nroot7": lambda x: sympy.Pow(x,1/7), 
                          "nroot11": lambda x: sympy.Pow(x,1/11)},_

Unfortunately, I've got the following error:

RuntimeError: <PyCall.jlwrap (in a Julia function called from Python)
JULIA: The operator `nroot5` is not well-defined over the real line, as it threw the error `UndefVarError` when evaluating the input -100.0. You can work around this by returning NaN for invalid inputs. For example, `safe_log(x::T) where {T} = x > 0 ? log(x) : T(NaN)`.

n-th root for the odd values should work for negative input. Please help me, what I am doing wrong?

Thank you in advance for your help!

With best regards,

Roman

**Comment by MilesCranmer** [comment, marked as answer]
Hi @romanszewczyk,

Thanks for the question. In PySR all operators need to handle any real-valued input. So because in Julia, custom powers require the base to be non-negative, you should instead write: `x >= 0 ? x^(1//5) : -(-x)^(1//5)` as the operator (the `//` creates a rational number). And for operators which are invalid for negatives, you can just return a NaN like `x >= 0 ? x^(1//4) : typeof(x)(NaN)`.

Cheers,
Miles

**Reply by romanszewczyk** [reply]
Hi Miles!

Thank you for your prompt answer! 

I changed to:

_unary_operators=["sqrt", "square", "cbrt", "cube", 
                     "inv(x)=1/x",
                     "nroot5(x)= x >= 0 ? x^(1//5) : -(-x)^(1//5)",
                     "nroot7(x)= x >= 0 ? x^(1//7) : -(-x)^(1//7)",
                     "nroot11(x)= x >= 0 ? x^(1//11) : -(-x)^(1//11)"],_

and now it works without problems!

I previously tried to use 5th and 7th powers:

_unary_operators=["sqrt", "square", "cbrt", "cube", 
                     "inv(x)=1/x",
                     "pow5(x)= x ^5",
                     "pow7(x)= x ^7",
                     "pow11(x)= x ^11"],_

and it didn't generated errors.

Should I change it also to:

_unary_operators=["sqrt", "square", "cbrt", "cube", 
                     "inv(x)=1/x",
                     "pow5(x)= x >= 0 ? x^5 : -(-x)^5",
                     "pow7(x)= x >= 0 ? x^7 : -(-x)^7",
                     "pow11(x)= x >= 0 ? x^11 : -(-x)^11"],_

or I can leave it as it was (if there is no fractional value in power)?

Thank you very much again for your help!

With best regards,

Roman

**Reply by MilesCranmer** [reply]
x^5 is fine as is. Just the non-integer powers need the conditional statement.

**Reply by romanszewczyk** [reply]
Hi!

Thank you again for your help and all your efforts!

I updated the code, and it seems that nroot5() and other nroot functions don't work with _turbo=True_ mode.

I received error:
JULIA: TypeError: non-boolean (VectorizationBase.Mask{4, UInt8}) used in boolean context

I switched _turbo=False_ and the problem disappeared. 

Is there any possibility to keep turbo mode with defined nroot() functions? 

Thank you very much again for your help!

With best regards!

Roman

**Reply by MilesCranmer** [reply]
Unfortunately turbo mode only works with operators that are capable of being turned into SIMD kernels. Branching statements tend not to be automatically/efficiently vectorizable.

In this case you could try something much more advanced… (needs some Julia background to understand what’s going on) say an operator definition with multiple dispatch like
```python
"""nroot5(x::Real)= x >= 0 ? x^(1//5) : -(-x)^(1//5)
nroot5(x) = x ^ (1/5)"""
```

make sure they’re on multiple lines. But I’m not sure this will work. Some operators have vectorized versions that have the NaN return built-in. Thus, declaring a generic version on the second line would cause the SIMD tracing to just call the normal operator, thus allowing the vectorization to do the NaN checking, and there’s no need for us to add it.

**Reply by MilesCranmer** [reply]
Or even easier is just

```julia
nroot5(x) = x isa Real ? (x >= 0 ? x^(1//5) : -(-x)^(1//5)) : x ^ (1/5)
```

so you don’t need the multi-line version.

**Reply by MilesCranmer** [reply]
Although actually I’m not sure there are vectorized versions for generic `^` which can work with negative numbers. So you might just need to avoid turbo mode.

**Reply by romanszewczyk** [reply]
Hi!

Thank you very much again for your guidelines.
I tried both variants and in both cases, I had the same error: 
JULIA: TypeError: non-boolean (VectorizationBase.Mask{4, UInt8}) used in boolean context

Finally, I set _turbo=False_  and used formula: _"nroot5(x) = x >= 0 ? x^(1//5) : -(-x)^(1//5)",_
It will be slower but more reliable.

Thank you very much again for all your support!

With best regards!
Roman
